cli/src/commands/parallelize.rs

```diff
@@ -0,0 +1,192 @@
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::io::Write;
+use std::rc::Rc;
+
+use indexmap::IndexSet;
+use itertools::Itertools;
+use jj_lib::backend::CommitId;
+use jj_lib::commit::Commit;
+use jj_lib::repo::Repo;
+use jj_lib::revset::{RevsetExpression, RevsetIteratorExt};
+use tracing::instrument;
+
+use crate::cli_util::{short_commit_hash, CommandHelper, RevisionArg};
+use crate::command_error::{user_error, CommandError};
+use crate::commands::rebase::rebase_descendants;
+use crate::ui::Ui;
+
+/// Parallelize revisions by making them siblings
+///
+/// The set of target commits being parallelized must have a single head and
+/// root commit (not to be confused with the repo root), otherwise the command
+/// will fail. Each of the target commits is rebased onto the parents of the
+/// root commit. The children of the head commit are rebased onto the target
+/// commits.
+///
+/// Example usage:
+/// ```text
+/// jj log
+/// ◉   3
+/// ◉   2
+/// ◉   1
+/// ◉
+/// jj parallelize 1::2
+/// jj log
+/// ◉     3
+/// ├─╮
+/// │ ◉   2
+/// ◉ │   1
+/// ├─╯
+/// ◉
+/// ```
+#[derive(clap::Args, Clone, Debug)]
+#[command(verbatim_doc_comment, visible_aliases=&["para"])]
+pub(crate) struct ParallelizeArgs {
+    revisions: Vec<RevisionArg>,
+}
+
+#[instrument(skip_all)]
+pub(crate) fn cmd_parallelize(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    args: &ParallelizeArgs,
+) -> Result<(), CommandError> {
+    let mut workspace_command = command.workspace_helper(ui)?;
+    let mut tx = workspace_command.start_transaction();
+    // The target commits are the commits being parallelized.
+    let target_commits: Vec<CommitId> = tx
+        .base_workspace_helper()
+        .parse_union_revsets(&args.revisions)?
+        .evaluate()?
+        .iter()
+        // We want parents before children, so we need to reverse the order.
+        .reversed()
+        .collect();
+    if target_commits.len() < 2 {
+        return Ok(writeln!(ui.status(), "Nothing changed.")?);
+    }
+    let target_revset = RevsetExpression::commits(target_commits.clone());
+    let target_head = get_head(&target_revset, tx.repo())?;
+    let target_root = get_root(&target_revset, tx.repo())?;
+    let connected_length = RevsetExpression::commits(vec![target_head, target_root.clone()])
+        .connected()
+        .evaluate_programmatic(tx.repo())?
+        .iter()
+        .count();
+    if connected_length != target_commits.len() {
+        return Err(user_error(
+            "Cannot parallelize since the target revisions are not connected.",
+        ));
+    }
```

Comment:
Since the "is linear and connect" condition is so simple, should we write it like this?
```
let target_root = target_commits[0];
if target_root.parent_ids().len() > 1 {
  // Error
}
for i in 1..target_commits.len() {
  if target_commits[i].parent_ids() != &[target_commits[i-1].id().clone()] {
    // Error
  }
}
```

Actually, will the current code catch a case where the head commit is a merge commit but you only include a linear set of commits from one side?

Reply:
> Actually, will the current code catch a case where the head commit is a merge commit but you only include a linear set of commits from one side?
I think it's more than that - there seem to be quite a few conditions that must be met. Let's define "tips" of a revision set as the commits in a revision set who have at least one child not in the revision set or who have no children. (This is not the "head" described in the documentation comment above; a "head" is not in the revision set.)
- single root commit that's *not* in the revision set
- if a "tip" has a child, that child commit must have all tips as its parents
- the revision set must have all commits between all "tips" and the root

For this reason, it might be easier to ask the user to specify the root and the tips (and specify zero tips if the user wants to parallelize all descendants of the root). If we do this, I think that the input restrictions will be clearer to the user.

Reply:
I think I'm against making the user specify the root and tip. That sounds cumbersome, and I think a common scenario for this function is to parallelize a string of commits from some root commit all the way to a head commit with no children. In that the current design makes it as simple as running `jj parallelize xyz::` or even `jj parallelize @::` depending on where `@` is sitting. Allowing the user to leave the tips out helps, but I suppose they'd still need to pass `--root` for consistency.

> if a "tip" has a child, that child commit must have all tips as its parents

I'm having a little trouble understanding your definition of "tips" and how that plus the rules you listed (specifically the quoted one) differ from the current rules for the function. Can you maybe give an example? Would any of the current test cases break?

Also, I think I am using the standard definition of root and head from the revset docs. In case that helps clear up any ambiguity:

- roots(x): Commits in x that are not descendants of other commits in x. Note that this is different from [Mercurial's](https://repo.mercurial-scm.org/hg/help/revsets) roots(x) function, which is equivalent to x ~ x+.
- heads(x): Commits in x that are not ancestors of other commits in x. Note that this is different from [Mercurial's](https://repo.mercurial-scm.org/hg/help/revsets) heads(x) function, which is equivalent to x ~ x-.

Reply:
> Actually, will the current code catch a case where the head commit is a merge commit but you only include a linear set of commits from one side?

No, it doesn't. I added some tests and the command allows that scenario and breaks the merge commit. I think we have a problem with this in general actually because the results also look wrong when an interior target is a merge. We can either fix how the commits are rebased, or return in a error in this scenario.

The new tests are `test_parallelize_head_is_a_merge` and `test_parallelize_interior_target_is_a_merge`.

Reply:
Thanks for the reference to the revset docs.

> I'm having a little trouble understanding your definition of "tips" and how that plus the rules you listed (specifically the quoted one) differ from the current rules for the function. Can you maybe give an example? Would any of the current test cases break?

The current rules for the function allow the case where "the head commit is a merge commit but you only include a linear set of commits from one side" (which breaks the function, I believe). But my definitions forbid that.

      d
     / \
    c   y
    |   |
    b   x
     \ /
      a

Let's say only a..d are in the set. In my definition, "a" is also a tip (since it has a child not in the set); the tips are "d" and "a". And "a" has a child "x", so "x" must have as parents "d" and "a", but this is not true, so this set a..d is invalid. I now realize that my definition of "tip" is the same as the Mercurial "heads".

After some more thought, I think that the rules can be further simplified into:
- Every commit in `x ~ x+` (the Mercurial roots) must have the same parents.
- Every commit in `x ~ x-` (the Mercurial heads) must have the same children.

(Note that if there is only one Mercurial root, the condition is satisfied. Likewise for one Mercurial head.)

Reply:
I created your example in an empty repo and as far as I can tell `roots(a..d)` is equivalent to `(a..d) ~(a..d)+`. Same for heads. Is there supposed to be a difference? The `a..d` revset includes everything except `a`.

In your example `a..d` has two roots, `b` and `x`, so the current rules already disallow it. Maybe you meant `b::c`, but even then the Mercurial and jj roots and heads are equivalent.

The "head is a merge commit" issue appears to be real, but I'm not sure these rules solve it. Perhaps all we need to do is check that all of the parents of the head commit are in the target set being parallelized. Between that check, and the check that there is only one root, I think we'll prevent `d` in your example from being parallelized unless the user parallelizes all of `[d, c, b, y, x]`.

Reply:
I pushed a new version with a bunch of additional tests. This strange merge situation is no longer allowed. The new rule is that the parents of every target commit __except for the root__ must also be in the target set. This means that in the example above you can only parallelize all of `[d, c, b, y, x]`.

Reply:
Ah, I think I got some of the revset notation wrong. But...

> the parents of every target commit except for the root must also be in the target set

This is something like what I was trying to get at (and I think you need a similar rule "children of every target commit except for heads must also be in the target set, but I might be wrong) and you've done much more thinking on this than I have, so I'll leave it in your hands.

Reply:
I think we could add that rule, but it's not strictly necessary. If the user doesn't follow it the resulting history can be complex. So it's a matter of whether we want to allow that or not. Those types of complex histories probably aren't very useful (I wouldn't find them to be very useful), but maybe someone does?

We can think about it some more before deciding. We could even hide the complex rebasing logic behind a flag. If the flag isn't specified, parallelize can enforce a stricter set of rules about the target revisions.